go-controller/pkg/ovn/address_set.go

```diff
@@ -0,0 +1,259 @@
+	oc.addressSets[as.name] = as
+}
+
+// deleteAddressSet removes the address set from OVN and if the address set
+// is known to the Controller, removes it from the controller's address set map
+func (oc *Controller) deleteAddressSet(name string) {
+	oc.addressSetMutex.Lock()
+	defer oc.addressSetMutex.Unlock()
+	if as, ok := oc.addressSets[name]; ok {
+		as.Destroy()
+		delete(oc.addressSets, name)
+	} else {
+		destroyAddressSet(hashedAddressSet(name))
+	}
+}
+
+type addressSet struct {
+	sync.Mutex
+
+	name      string
+	hashName  string
+	ports     map[string]net.IP // logical port :: pod IP
+	destroyed bool
+}
+
+// hash the provided input to make it a valid addressSet name.
+func hashedAddressSet(s string) string {
+	return hashForOVN(s)
+}
+
+type addressSetIterFn func(name, namespace, suffix string)
+
+// forEachAddressSetUnhashedName will pass the unhashedName, namespaceName and
+// the first suffix in the name to the 'iteratorFn' for every address_set in
+// OVN. (Each unhashed name for an addressSet can be of the form
+// namespaceName.suffix1.suffix2. .suffixN)
+func forEachAddressSetUnhashedName(iteratorFn addressSetIterFn) error {
+	output, stderr, err := util.RunOVNNbctl("--data=bare", "--no-heading",
+		"--columns=external_ids", "find", "address_set")
+	if err != nil {
+		klog.Errorf("Error in obtaining list of address sets from OVN: "+
+			"stdout: %q, stderr: %q err: %v", output, stderr, err)
+		return err
+	}
+	for _, addrSet := range strings.Fields(output) {
+		if !strings.HasPrefix(addrSet, "name=") {
+			continue
+		}
+		addrSetName := addrSet[5:]
+		names := strings.Split(addrSetName, ".")
+		addrSetNamespace := names[0]
+		nameSuffix := ""
+		if len(names) >= 2 {
+			nameSuffix = names[1]
+		}
+		iteratorFn(addrSetName, addrSetNamespace, nameSuffix)
+	}
+	return nil
+}
+
+func NewAddressSet(name string, ports map[string]net.IP) *addressSet {
+	if ports == nil {
+		ports = make(map[string]net.IP)
+	}
+	as := &addressSet{
+		name:     name,
+		hashName: hashForOVN(name),
+		ports:    ports,
+	}
+
+	klog.V(5).Infof("New(%s/%s) with %v", as.hashName, as.name, ports)
+
+	addressSet, stderr, err := util.RunOVNNbctl("--data=bare",
+		"--no-heading", "--columns=_uuid", "find", "address_set",
+		fmt.Sprintf("name=%s", as.hashName))
+	if err != nil {
+		klog.Errorf("find failed to get address set, stderr: %q (%v)",
+			stderr, err)
+	}
+
+	// addressSet has already been created in the database and nothing to set.
+	if addressSet != "" {
+		as.setOrClear()
+	} else {
+		// addressSet has not been created yet. Create it.
+		args := []string{
+			"create",
+			"address_set",
+			"name=" + as.hashName,
+			"external-ids:name=" + as.name,
+		}
+		joinedIPs, haveAny := as.joinIPs()
+		if haveAny {
+			args = append(args, "addresses="+joinedIPs)
+		}
+		_, stderr, err = util.RunOVNNbctl(args...)
+		if err != nil {
+			klog.Errorf("failed to create address_set %s, stderr: %q (%v)",
+				as.name, stderr, err)
```

Comment:
we really ought to be returning an error here and causing the failure to propagate up (re #1279)